src/researchtikpy/get_videos_hashtag.py

```diff
@@ -87,19 +79,94 @@ def get_videos_query(query: dict, access_token: str, start_date: str, end_date:
             "max_count": max_count
         }
 
-        response = requests.post(url_with_fields, headers=headers, json=query_body)
-        if response.status_code == 200:
-            data = response.json().get("data", {})
-            videos = data.get("videos", [])
-            collected_videos.extend(videos)
-            if not data.get("has_more", False) or len(collected_videos) >= total_max_count:
+        for response in iter_responses(query_body, access_token):
+            if_needed_log_failures_and_wait(response)
+            if response.status_code == 200:
+                videos: list[dict] = response.json()["data"]["videos"]
+                print(f"Received {len(videos)} videos.")
+                collected_videos.extend(videos)
+            if len(collected_videos) >= total_max_count:
                 break
-        elif response.status_code == 429:
-            print("Rate limit exceeded. Pausing before retrying...")
-            time.sleep(rate_limit_pause)
-        else:
-            raise ValueError(f"status_code={response.status_code}. {response.json()}")
 
         start_date_dt = current_end_date + timedelta(days=1)
 
     return pd.DataFrame(collected_videos[:total_max_count])
+
+
+def post_query(full_query: dict, access_token: str) -> requests.Response:
+    """The full query includes e.g. 'max_count', 'search_id' and 'cursor' fields."""
+    assert isinstance(access_token, str), "access_token must be a string!"
+    endpoint = "https://open.tiktokapis.com/v2/research/video/query/"
+    fields = "id,video_description,create_time,region_code,share_count,view_count,like_count,comment_count,music_id,hashtag_names,username,effect_ids,playlist_id,voice_to_text"
+    url_with_fields = f"{endpoint}?fields={fields}"
+    headers = {
+        "Content-Type": "application/json",
+        "Authorization": f"Bearer {access_token}"
+    }
+    print(f"Calling TikTok API with data={full_query}")
+    return requests.post(url_with_fields, headers=headers, json=full_query)
+
+
+def iter_responses(query_body: dict, access_token: str) -> Iterator[requests.Response]:
+    """
+    Creates an iterator that uses the cursor-based pagination to request all videos sequentially.
+    If query_body['is_random'] is True, the iterator will not use cursor pagination.
+    Each element yielded is an http response from the API.
+    """
+    search_id = query_body.get("search_id", None)
+    cursor = query_body.get("cursor", 0)
+    is_sequential = not bool(query_body.get("is_random", False))
+
+    while True:
+        full_query = query_body | dict(search_id=search_id, cursor=cursor)
+        response = post_query(full_query=full_query, access_token=access_token)
+        yield response
+        if response.status_code == 200:
```

Comment:
If the API happened to return invalid JSON, could we catch and retry here?

Reply:
Sounds reasonable. There are several ways to go about this, do you have anything specific in mind?